android/app/src/main/java/org/cryptimeleon/incentive/app/domain/usecase/PayAndRedeemUseCase.kt

```diff
@@ -0,0 +1,239 @@
+import kotlinx.coroutines.flow.flow
+import org.cryptimeleon.craco.sig.sps.eq.SPSEQSignature
+import org.cryptimeleon.incentive.app.domain.IBasketRepository
+import org.cryptimeleon.incentive.app.domain.ICryptoRepository
+import org.cryptimeleon.incentive.app.domain.IPromotionRepository
+import org.cryptimeleon.incentive.app.domain.model.BulkRequestDto
+import org.cryptimeleon.incentive.app.domain.model.Earn
+import org.cryptimeleon.incentive.app.domain.model.EarnRequestData
+import org.cryptimeleon.incentive.app.domain.model.PromotionUserUpdateChoice
+import org.cryptimeleon.incentive.app.domain.model.SpendRequestData
+import org.cryptimeleon.incentive.app.domain.model.ZKP
+import org.cryptimeleon.incentive.app.util.toBigIntVector
+import org.cryptimeleon.incentive.crypto.IncentiveSystem
+import org.cryptimeleon.incentive.crypto.model.EarnRequest
+import org.cryptimeleon.incentive.crypto.model.SpendRequest
+import org.cryptimeleon.incentive.crypto.model.SpendResponse
+import org.cryptimeleon.incentive.crypto.model.Token
+import org.cryptimeleon.incentive.promotion.Promotion
+import org.cryptimeleon.incentive.promotion.ZkpTokenUpdateMetadata
+import org.cryptimeleon.math.serialization.RepresentableRepresentation
+import org.cryptimeleon.math.serialization.converter.JSONConverter
+import org.cryptimeleon.math.structures.cartesian.Vector
+import timber.log.Timber
+import java.math.BigInteger
+
+class PayAndRedeemUseCase(
+    private val promotionRepository: IPromotionRepository,
+    private val cryptoRepository: ICryptoRepository,
+    private val basketRepository: IBasketRepository,
+) {
+    private val jsonConverter = JSONConverter()
+    private val analyzeUserTokenUpdatesUseCase =
+        AnalyzeUserTokenUpdatesUseCase(promotionRepository, cryptoRepository, basketRepository)
+
+    /**
+     * Send all earn and spend requests for a basket, pay and retrieve updated tokens
+     *
+     * This process will look as follows:
+     * 0. Check user choices present for all promotions (or default to None?)
+     * 1. Prepare all requests for Earn and ZKP (send as a batch?)
+     * 2. Pay basket
+     * 3. Obtain results to the Earn/ZKP requests
+     * 4. Update tokens
+     * 5. Clear basket and user choices
+     */
+    suspend operator fun invoke(): Flow<PayAndRedeemState> =
+        flow {
+            // Some setup
+            val userTokenUpdates: List<PromotionUserUpdateChoice> =
+                analyzeUserTokenUpdatesUseCase().first()
+            val basketId = basketRepository.basket.first()!!.basketId
+            val basket = basketRepository.basket.first()!!
+            val promotionParameters = promotionRepository.promotions.first()
+            val cryptoMaterial = cryptoRepository.cryptoMaterial.first()!!
+            val tokens = cryptoRepository.tokens.first()
+            val incentiveSystem = IncentiveSystem(cryptoMaterial.pp)
+
+            // TODO some checks on the basket and user choices
```

Comment:
One of the things "put to issues so branch does not get too big"?

Reply:
https://github.com/cryptimeleon/incentive-system/issues/98